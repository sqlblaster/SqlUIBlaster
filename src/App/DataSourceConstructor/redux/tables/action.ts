import { Action } from 'src/components/query-builder/store/models/action-types';
import { Table } from '../../schemas';

export type TablesAction = Action<'SET_TABLES', { tables: Table[] }>;

export const setTables = (tables: Table[]): TablesAction => ({
  type: 'SET_TABLES',
  tables
});
