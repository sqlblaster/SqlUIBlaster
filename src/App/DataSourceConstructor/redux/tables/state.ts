import { Table } from '../../schemas';

export class TablesState {
  public tables: Table[] = [];
}
