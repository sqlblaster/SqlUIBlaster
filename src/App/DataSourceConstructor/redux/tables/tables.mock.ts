import { Column, Table } from '../../schemas';

export const tablesMock = [
  new Table({
    name: 'Orders',
    columns: [
      new Column({
        name: 'ID',
        type: 'primary'
      }),
      new Column({
        name: 'Details',
        type: 'string',
        enums: ['Detail1', 'Detail2']
      }),
      new Column({
        name: 'Created At',
        type: 'Date'
      }),
      new Column({
        name: 'Quantity',
        type: 'number'
      }),
      new Column({
        name: 'Paid',
        type: 'boolean'
      }),
      new Column({
        name: 'ProductID',
        type: 'string',
        foreignTableName: 'Products'
      }),
      new Column({
        name: 'UserID',
        type: 'string',
        foreignTableName: 'Users'
      })
    ]
  }),
  new Table({
    name: 'Users',
    columns: [
      new Column({
        name: 'ID',
        type: 'primary'
      }),
      new Column({
        name: 'AddressID',
        type: 'number',
        foreignTableName: 'Addresses'
      }),
      new Column({
        name: 'Username',
        type: 'string'
      }),
      new Column({
        name: 'AdminID',
        type: 'string',
        foreignTableName: 'Users'
      })
    ]
  }),
  new Table({
    name: 'Products',
    columns: [
      new Column({
        name: 'ID',
        type: 'primary'
      }),
      new Column({
        name: 'Product name',
        type: 'string'
      }),
      new Column({
        name: 'Available amount',
        type: 'number'
      }),
      new Column({
        name: 'AddedBy',
        type: 'string',
        foreignTableName: 'Users'
      }),
      new Column({
        name: 'CurrencyID',
        type: 'string',
        foreignTableName: 'Currency'
      }),
      new Column({
        name: 'LocalCurrencyID',
        type: 'string',
        foreignTableName: 'Currency'
      })
    ]
  }),
  new Table({
    name: 'Addresses',
    columns: [
      new Column({
        name: 'ID',
        type: 'primary'
      }),
      new Column({
        name: 'StreetID',
        type: 'string',
        foreignTableName: 'Streets'
      })
    ]
  }),
  new Table({
    name: 'Streets',
    columns: [
      new Column({
        name: 'ID',
        type: 'primary'
      }),
      new Column({
        name: 'Street name',
        type: 'string'
      })
    ]
  }),
  new Table({
    name: 'Currency',
    primaryKeyColumn: 'ID',
    columns: [
      new Column({
        name: 'ID',
        type: 'primary'
      }),
      new Column({
        name: 'name',
        type: 'string'
      })
    ]
  })
];
