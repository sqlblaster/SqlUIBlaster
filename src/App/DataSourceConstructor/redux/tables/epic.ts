import { ofType } from 'redux-observable';
import { of } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { raiseGlobalError } from 'src/store/global-error/action';
import { Epic } from 'src/store/models/Epic';
import { SetDatabaseSchemaAction } from '../database-schema/action';
import { setTables } from './action';

export const extractTablesEpic: Epic<SetDatabaseSchemaAction> = (
  action$,
  state$
) =>
  action$.pipe(
    ofType<SetDatabaseSchemaAction>(
      'SET_DATABASE_SCHEMA'
    ),
    map(({ databaseSchema }) =>
      databaseSchema
        ? setTables(databaseSchema.tables)
        : raiseGlobalError('Couldn\'t fetch database schema')
    ),
    catchError((err) => {
      console.error(err && err.xhr && err.xhr.response);

      return of(raiseGlobalError('Couldn\'t fetch database schema'));
    })
  );
