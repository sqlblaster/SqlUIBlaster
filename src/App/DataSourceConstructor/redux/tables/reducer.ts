import { Reducer } from 'redux';
import { Table } from '../../schemas';
import { TablesAction } from './action';

export const tables: Reducer<Table[], TablesAction> = (
  state = [],
  { type, tables: newTables }
) => {
  return type === 'SET_TABLES' ? newTables : state;
};
