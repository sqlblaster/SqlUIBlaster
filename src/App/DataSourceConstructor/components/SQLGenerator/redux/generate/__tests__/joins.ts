import { tablesMock } from 'src/App/DataSourceConstructor/redux/tables/tables.mock';
import { generateSQLQuery, SQLGeneratorStateProps } from '..';
import { zeroLvlGroupings } from './utils.joins/0-lvl-groupings';
import { oneLvlGroupings } from './utils.joins/1-lvl-groupings';
import { twoLvlGroupings } from './utils.joins/2-lvl-groupings';
import { threeLvlGroupings } from './utils.joins/3-lvl-groupings';
import {
  selfRef1LvlGroupings,
  selfRef2LvlGroupings,
  selfRef3LvlGroupings
} from './utils.joins/self-referencing-groupings';
import { similarForeignGroupings } from './utils.joins/similar-foreign-groupings';

const state: SQLGeneratorStateProps = {
  selectedTable: tablesMock[0],
  customColumns: [],
  filters: [],
  aggregations: [],
  groupings: [],
  orders: [],
  rowLimit: null,
  tables: tablesMock
};

describe('JOINs', () => {
  it('should be rendered nothing when there is only selected table', () => {
    expect(generateSQLQuery(state)).toMatchSnapshot();
  });

  it('should be rendered nothing when there are 0 lvl column', () => {
    expect(
      generateSQLQuery({ ...state, groupings: [zeroLvlGroupings[0]] })
    ).toMatchSnapshot();
  });

  it('should be rendered nothing when there are 0 lvl columns', () => {
    expect(
      generateSQLQuery({ ...state, groupings: zeroLvlGroupings })
    ).toMatchSnapshot();
  });

  it(`should be rendered properly when there are
      two 1 lvl columns on the same foreign table`, () => {
    expect(
      generateSQLQuery({ ...state, groupings: similarForeignGroupings })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 1 lvl column', () => {
    expect(
      generateSQLQuery({ ...state, groupings: [oneLvlGroupings[0]] })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 1 lvl columns', () => {
    expect(
      generateSQLQuery({ ...state, groupings: oneLvlGroupings })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 2 lvl column', () => {
    expect(
      generateSQLQuery({ ...state, groupings: [twoLvlGroupings[0]] })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 2 lvl columns', () => {
    expect(
      generateSQLQuery({ ...state, groupings: twoLvlGroupings })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 1, 2 lvl columns', () => {
    expect(
      generateSQLQuery({
        ...state,
        groupings: [oneLvlGroupings[0], twoLvlGroupings[0]]
      })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 3 lvl column', () => {
    expect(
      generateSQLQuery({ ...state, groupings: [threeLvlGroupings[0]] })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 3 lvl columns', () => {
    expect(
      generateSQLQuery({ ...state, groupings: threeLvlGroupings })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 2, 3 lvl columns', () => {
    expect(
      generateSQLQuery({
        ...state,
        groupings: [twoLvlGroupings[0], threeLvlGroupings[0]]
      })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 1, 2, 3 lvl columns', () => {
    expect(
      generateSQLQuery({
        ...state,
        groupings: [
          oneLvlGroupings[0],
          twoLvlGroupings[0],
          threeLvlGroupings[0]
        ]
      })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 1 lvl self-referencing column', () => {
    expect(
      generateSQLQuery({ ...state, groupings: [selfRef1LvlGroupings[0]] })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 1 lvl self-referencing columns', () => {
    expect(
      generateSQLQuery({ ...state, groupings: selfRef1LvlGroupings })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 2 lvl self-referencing column', () => {
    expect(
      generateSQLQuery({ ...state, groupings: [selfRef2LvlGroupings[0]] })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 2 lvl self-referencing columns', () => {
    expect(
      generateSQLQuery({ ...state, groupings: selfRef2LvlGroupings })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 3 lvl self-referencing column', () => {
    expect(
      generateSQLQuery({ ...state, groupings: [selfRef3LvlGroupings[0]] })
    ).toMatchSnapshot();
  });

  it('should be rendered properly when there are 3 lvl self-referencing columns', () => {
    expect(
      generateSQLQuery({ ...state, groupings: selfRef3LvlGroupings })
    ).toMatchSnapshot();
  });

  it(`should be rendered properly when there are 1, 2, 3 lvl
      and 3 lvl self-referencing column`, () => {
    expect(
      generateSQLQuery({
        ...state,
        groupings: [
          selfRef1LvlGroupings[0],
          selfRef2LvlGroupings[0],
          selfRef3LvlGroupings[0]
        ]
      })
    ).toMatchSnapshot();
  });
});
