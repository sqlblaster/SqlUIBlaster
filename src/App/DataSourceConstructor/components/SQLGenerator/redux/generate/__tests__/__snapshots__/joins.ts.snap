// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JOINs should be rendered nothing when there are 0 lvl column 1`] = `
"
    SELECT
      \\"orders1\\".\\"ID\\" AS \\"ID (orders1)\\"
    FROM \\"orders\\" \\"orders1\\"
    GROUP BY \\"orders1\\".\\"ID\\"
    ORDER BY \\"orders1\\".\\"ID\\" ASC"
`;

exports[`JOINs should be rendered nothing when there are 0 lvl columns 1`] = `
"
    SELECT
      \\"orders1\\".\\"ID\\" AS \\"ID (orders1)\\",\\"orders1\\".\\"Created At\\" AS \\"Created At (orders1)\\",\\"orders1\\".\\"Details\\" AS \\"Details (orders1)\\",\\"orders1\\".\\"Quantity\\" AS \\"Quantity (orders1)\\"
    FROM \\"orders\\" \\"orders1\\"
    GROUP BY \\"orders1\\".\\"ID\\",\\"orders1\\".\\"Created At\\",\\"orders1\\".\\"Details\\",\\"orders1\\".\\"Quantity\\"
    ORDER BY \\"orders1\\".\\"ID\\" ASC,\\"orders1\\".\\"Created At\\" ASC,\\"orders1\\".\\"Details\\" ASC,\\"orders1\\".\\"Quantity\\" ASC"
`;

exports[`JOINs should be rendered nothing when there is only selected table 1`] = `
"
    SELECT
      *
    FROM \\"orders\\" "
`;

exports[`JOINs should be rendered properly when there are
      two 1 lvl columns on the same foreign table 1`] = `
"
    SELECT
      \\"currency1\\".\\"name\\" AS \\"name (currency1)\\",\\"currency2\\".\\"ID\\" AS \\"ID (currency2)\\"
    FROM \\"orders\\" \\"products1\\"
  LEFT JOIN \\"currency\\" \\"currency1\\"
  ON \\"products1\\".\\"CurrencyID\\" = \\"currency1\\".\\"ID\\"
  LEFT JOIN \\"currency\\" \\"currency2\\"
  ON \\"products1\\".\\"LocalCurrencyID\\" = \\"currency2\\".\\"ID\\"
    GROUP BY \\"currency1\\".\\"name\\",\\"currency2\\".\\"ID\\"
    ORDER BY \\"currency1\\".\\"name\\" ASC,\\"currency2\\".\\"ID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 1 lvl column 1`] = `
"
    SELECT
      \\"products1\\".\\"ID\\" AS \\"ID (products1)\\"
    FROM \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\"
  ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
    GROUP BY \\"products1\\".\\"ID\\"
    ORDER BY \\"products1\\".\\"ID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 1 lvl columns 1`] = `
"
    SELECT
      \\"products1\\".\\"ID\\" AS \\"ID (products1)\\",\\"users1\\".\\"Username\\" AS \\"Username (users1)\\",\\"products1\\".\\"Product name\\" AS \\"Product name (products1)\\"
    FROM \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\"
  ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users1\\"
  ON \\"orders1\\".\\"UserID\\" = \\"users1\\".\\"ID\\"
    GROUP BY \\"products1\\".\\"ID\\",\\"users1\\".\\"Username\\",\\"products1\\".\\"Product name\\"
    ORDER BY \\"products1\\".\\"ID\\" ASC,\\"users1\\".\\"Username\\" ASC,\\"products1\\".\\"Product name\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 1 lvl self-referencing column 1`] = `
"
    SELECT
      \\"users2\\".\\"Username\\" AS \\"Username (users2)\\"
    FROM \\"orders\\" \\"users1\\"
  LEFT JOIN \\"users\\" \\"users2\\"
  ON \\"users1\\".\\"AdminID\\" = \\"users2\\".\\"ID\\"
    GROUP BY \\"users2\\".\\"Username\\"
    ORDER BY \\"users2\\".\\"Username\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 1 lvl self-referencing columns 1`] = `
"
    SELECT
      \\"users2\\".\\"Username\\" AS \\"Username (users2)\\",\\"users2\\".\\"ID\\" AS \\"ID (users2)\\",\\"users2\\".\\"AddressID\\" AS \\"AddressID (users2)\\"
    FROM \\"orders\\" \\"users1\\"
  LEFT JOIN \\"users\\" \\"users2\\"
  ON \\"users1\\".\\"AdminID\\" = \\"users2\\".\\"ID\\"
    GROUP BY \\"users2\\".\\"Username\\",\\"users2\\".\\"ID\\",\\"users2\\".\\"AddressID\\"
    ORDER BY \\"users2\\".\\"Username\\" ASC,\\"users2\\".\\"ID\\" ASC,\\"users2\\".\\"AddressID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 1, 2 lvl columns 1`] = `
"
    SELECT
      \\"products1\\".\\"ID\\" AS \\"ID (products1)\\",\\"users1\\".\\"ID\\" AS \\"ID (users1)\\"
    FROM \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\"
  ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users1\\"
  ON \\"products1\\".\\"AddedBy\\" = \\"users1\\".\\"ID\\"
    GROUP BY \\"products1\\".\\"ID\\",\\"users1\\".\\"ID\\"
    ORDER BY \\"products1\\".\\"ID\\" ASC,\\"users1\\".\\"ID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 1, 2, 3 lvl
      and 3 lvl self-referencing column 1`] = `
"
    SELECT
      \\"users2\\".\\"Username\\" AS \\"Username (users2)\\",\\"users3\\".\\"ID\\" AS \\"ID (users3)\\",\\"users4\\".\\"ID\\" AS \\"ID (users4)\\"
    FROM \\"orders\\" \\"users1\\"
  LEFT JOIN \\"users\\" \\"users2\\"
  ON \\"users1\\".\\"AdminID\\" = \\"users2\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users3\\"
  ON \\"users2\\".\\"AdminID\\" = \\"users3\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users4\\"
  ON \\"users3\\".\\"AdminID\\" = \\"users4\\".\\"ID\\"
    GROUP BY \\"users2\\".\\"Username\\",\\"users3\\".\\"ID\\",\\"users4\\".\\"ID\\"
    ORDER BY \\"users2\\".\\"Username\\" ASC,\\"users3\\".\\"ID\\" ASC,\\"users4\\".\\"ID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 1, 2, 3 lvl columns 1`] = `
"
    SELECT
      \\"products1\\".\\"ID\\" AS \\"ID (products1)\\",\\"users1\\".\\"ID\\" AS \\"ID (users1)\\",\\"addresses1\\".\\"ID\\" AS \\"ID (addresses1)\\"
    FROM \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\"
  ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users1\\"
  ON \\"products1\\".\\"AddedBy\\" = \\"users1\\".\\"ID\\"
  LEFT JOIN \\"addresses\\" \\"addresses1\\"
  ON \\"users1\\".\\"AddressID\\" = \\"addresses1\\".\\"ID\\"
    GROUP BY \\"products1\\".\\"ID\\",\\"users1\\".\\"ID\\",\\"addresses1\\".\\"ID\\"
    ORDER BY \\"products1\\".\\"ID\\" ASC,\\"users1\\".\\"ID\\" ASC,\\"addresses1\\".\\"ID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 2 lvl column 1`] = `
"
    SELECT
      \\"users1\\".\\"ID\\" AS \\"ID (users1)\\"
    FROM \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\"
  ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users1\\"
  ON \\"products1\\".\\"AddedBy\\" = \\"users1\\".\\"ID\\"
    GROUP BY \\"users1\\".\\"ID\\"
    ORDER BY \\"users1\\".\\"ID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 2 lvl columns 1`] = `
"
    SELECT
      \\"users1\\".\\"ID\\" AS \\"ID (users1)\\",\\"users1\\".\\"Username\\" AS \\"Username (users1)\\",\\"addresses1\\".\\"ID\\" AS \\"ID (addresses1)\\"
    FROM \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\"
  ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users1\\"
  ON \\"products1\\".\\"AddedBy\\" = \\"users1\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users2\\"
  ON \\"orders1\\".\\"UserID\\" = \\"users2\\".\\"ID\\"
  LEFT JOIN \\"addresses\\" \\"addresses1\\"
  ON \\"users2\\".\\"AddressID\\" = \\"addresses1\\".\\"ID\\"
    GROUP BY \\"users1\\".\\"ID\\",\\"users1\\".\\"Username\\",\\"addresses1\\".\\"ID\\"
    ORDER BY \\"users1\\".\\"ID\\" ASC,\\"users1\\".\\"Username\\" ASC,\\"addresses1\\".\\"ID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 2 lvl self-referencing column 1`] = `
"
    SELECT
      \\"users3\\".\\"ID\\" AS \\"ID (users3)\\"
    FROM \\"orders\\" \\"users1\\"
  LEFT JOIN \\"users\\" \\"users2\\"
  ON \\"users1\\".\\"AdminID\\" = \\"users2\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users3\\"
  ON \\"users2\\".\\"AdminID\\" = \\"users3\\".\\"ID\\"
    GROUP BY \\"users3\\".\\"ID\\"
    ORDER BY \\"users3\\".\\"ID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 2 lvl self-referencing columns 1`] = `
"
    SELECT
      \\"users3\\".\\"ID\\" AS \\"ID (users3)\\",\\"users3\\".\\"Username\\" AS \\"Username (users3)\\",\\"users3\\".\\"AddressID\\" AS \\"AddressID (users3)\\"
    FROM \\"orders\\" \\"users1\\"
  LEFT JOIN \\"users\\" \\"users2\\"
  ON \\"users1\\".\\"AdminID\\" = \\"users2\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users3\\"
  ON \\"users2\\".\\"AdminID\\" = \\"users3\\".\\"ID\\"
    GROUP BY \\"users3\\".\\"ID\\",\\"users3\\".\\"Username\\",\\"users3\\".\\"AddressID\\"
    ORDER BY \\"users3\\".\\"ID\\" ASC,\\"users3\\".\\"Username\\" ASC,\\"users3\\".\\"AddressID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 2, 3 lvl columns 1`] = `
"
    SELECT
      \\"users1\\".\\"ID\\" AS \\"ID (users1)\\",\\"addresses1\\".\\"ID\\" AS \\"ID (addresses1)\\"
    FROM \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\"
  ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users1\\"
  ON \\"products1\\".\\"AddedBy\\" = \\"users1\\".\\"ID\\"
  LEFT JOIN \\"addresses\\" \\"addresses1\\"
  ON \\"users1\\".\\"AddressID\\" = \\"addresses1\\".\\"ID\\"
    GROUP BY \\"users1\\".\\"ID\\",\\"addresses1\\".\\"ID\\"
    ORDER BY \\"users1\\".\\"ID\\" ASC,\\"addresses1\\".\\"ID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 3 lvl column 1`] = `
"
    SELECT
      \\"addresses1\\".\\"ID\\" AS \\"ID (addresses1)\\"
    FROM \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\"
  ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users1\\"
  ON \\"products1\\".\\"AddedBy\\" = \\"users1\\".\\"ID\\"
  LEFT JOIN \\"addresses\\" \\"addresses1\\"
  ON \\"users1\\".\\"AddressID\\" = \\"addresses1\\".\\"ID\\"
    GROUP BY \\"addresses1\\".\\"ID\\"
    ORDER BY \\"addresses1\\".\\"ID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 3 lvl columns 1`] = `
"
    SELECT
      \\"addresses1\\".\\"ID\\" AS \\"ID (addresses1)\\",\\"users2\\".\\"Username\\" AS \\"Username (users2)\\",\\"streets1\\".\\"Street name\\" AS \\"Street name (streets1)\\"
    FROM \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\"
  ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users1\\"
  ON \\"products1\\".\\"AddedBy\\" = \\"users1\\".\\"ID\\"
  LEFT JOIN \\"addresses\\" \\"addresses1\\"
  ON \\"users1\\".\\"AddressID\\" = \\"addresses1\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users2\\"
  ON \\"users1\\".\\"AdminID\\" = \\"users2\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users3\\"
  ON \\"orders1\\".\\"UserID\\" = \\"users3\\".\\"ID\\"
  LEFT JOIN \\"addresses\\" \\"addresses2\\"
  ON \\"users3\\".\\"AddressID\\" = \\"addresses2\\".\\"ID\\"
  LEFT JOIN \\"streets\\" \\"streets1\\"
  ON \\"addresses2\\".\\"StreetID\\" = \\"streets1\\".\\"ID\\"
    GROUP BY \\"addresses1\\".\\"ID\\",\\"users2\\".\\"Username\\",\\"streets1\\".\\"Street name\\"
    ORDER BY \\"addresses1\\".\\"ID\\" ASC,\\"users2\\".\\"Username\\" ASC,\\"streets1\\".\\"Street name\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 3 lvl self-referencing column 1`] = `
"
    SELECT
      \\"users4\\".\\"ID\\" AS \\"ID (users4)\\"
    FROM \\"orders\\" \\"users1\\"
  LEFT JOIN \\"users\\" \\"users2\\"
  ON \\"users1\\".\\"AdminID\\" = \\"users2\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users3\\"
  ON \\"users2\\".\\"AdminID\\" = \\"users3\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users4\\"
  ON \\"users3\\".\\"AdminID\\" = \\"users4\\".\\"ID\\"
    GROUP BY \\"users4\\".\\"ID\\"
    ORDER BY \\"users4\\".\\"ID\\" ASC"
`;

exports[`JOINs should be rendered properly when there are 3 lvl self-referencing columns 1`] = `
"
    SELECT
      \\"users4\\".\\"ID\\" AS \\"ID (users4)\\",\\"users4\\".\\"Username\\" AS \\"Username (users4)\\",\\"users4\\".\\"AddressID\\" AS \\"AddressID (users4)\\"
    FROM \\"orders\\" \\"users1\\"
  LEFT JOIN \\"users\\" \\"users2\\"
  ON \\"users1\\".\\"AdminID\\" = \\"users2\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users3\\"
  ON \\"users2\\".\\"AdminID\\" = \\"users3\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users4\\"
  ON \\"users3\\".\\"AdminID\\" = \\"users4\\".\\"ID\\"
    GROUP BY \\"users4\\".\\"ID\\",\\"users4\\".\\"Username\\",\\"users4\\".\\"AddressID\\"
    ORDER BY \\"users4\\".\\"ID\\" ASC,\\"users4\\".\\"Username\\" ASC,\\"users4\\".\\"AddressID\\" ASC"
`;
